Immutable assets

Crisspl opened the pull request.

Crisspl committed:
Further handling of asset mutablity stuff

Crisspl committed:
Immutable assets basically done

However some TODOs are marked in the comments

devshgraphicsprogramming commented on include/irr/asset/IAsset.h:
```diff
@@ -167,25 +173,40 @@ class IAsset : virtual public core::IReferenceCounted
 
+		inline E_MUTABILITY getMutability() const { return m_mutability; }
+		inline bool isMutable() const { return getMutability() == EM_MUTABLE; }
+		inline bool canBeConvertedToDummy() const { return (getMutability()&EM_CPU_PERSISTENT) != EM_CPU_PERSISTENT; }
```
not sure I want these as public.

We should have a `setMutability`

Crisspl replied:
setting asset's mutability is exclusive for asset manager and it's already a friend of IAsset

devshgraphicsprogramming replied:
ah ok, but still the public functions here are a bit ugly, they should be protected.

Crisspl replied:
shouldnt user be able to query asset's mutability level?

devshgraphicsprogramming replied:
yes, but we shouldn't have multiple functions to query, just `getMutability` and the enum

obvs `canBeConvertedToDummy` can stay, cause it does more than checking an enum

devshgraphicsprogramming replied:
ok `isMutable` might actually be useful

devshgraphicsprogramming commented on include/irr/asset/IAsset.h:
```diff
@@ -167,25 +173,40 @@ class IAsset : virtual public core::IReferenceCounted
+		{
+			//we have to talk about it (TODO)
+			//if (isImmutable_debug())
+			//	return;
```
doesn't matter what you do here because of #451 which means that it should be the `SAssetBundle` that holds the metadata, not the `IAsset` (a future TODO along with #22 #23 which are still open because of a rare scheduling loophole)

devshgraphicsprogramming replied:
leave a comment pointing to the github issue.

devshgraphicsprogramming replied:
actually https://github.com/Devsh-Graphics-Programming/Nabla/issues/14

devshgraphicsprogramming replied:
leave a comment

devshgraphicsprogramming commented on include/irr/asset/ICPUBufferView.h:
```diff
@@ -44,11 +44,24 @@ class ICPUBufferView : public IBufferView<ICPUBuffer>, public IAsset
-		ICPUBuffer* getUnderlyingBuffer() { return m_buffer.get(); }
+		ICPUBuffer* getUnderlyingBuffer() 
+		{
+			return m_buffer.get(); 
```
forgot something?

devshgraphicsprogramming commented on include/irr/asset/ICPUDescriptorSet.h:
```diff
@@ -148,6 +151,10 @@ class ICPUDescriptorSet final : public IDescriptorSet<ICPUDescriptorSetLayout>,
 		{ 
+			//this must be commented-out, otherwise cpu2gpu doesnt have access to descriptors
+			//if (isImmutable_debug())
+			//	return core::SRange<SDescriptorInfo>{nullptr, nullptr};
```
does the cpu2gpu need non-`const` `SDescriptorInfo` ?

Crisspl replied:
need non-const descriptors

devshgraphicsprogramming replied:
but why?

devshgraphicsprogramming commented on include/irr/asset/ICPUDescriptorSet.h:
```diff
@@ -128,7 +128,10 @@ class ICPUDescriptorSet final : public IDescriptorSet<ICPUDescriptorSetLayout>,
 		_IRR_STATIC_INLINE_CONSTEXPR auto AssetType = ET_DESCRIPTOR_SET;
```
convertToDummyObject above this line does not check for mutability

Crisspl replied:
because it doesnt have to, it would change nothing

devshgraphicsprogramming commented on include/irr/asset/ICPUImage.h:
```diff
@@ -65,10 +65,14 @@ class ICPUImage final : public IImage, public IAsset
 
 		virtual bool validateCopies(const SImageCopy* pRegionsBegin, const SImageCopy* pRegionsEnd, const ICPUImage* src)
 		{
+			//TODO why is it not const method?
```
if it can be const, make it const

devshgraphicsprogramming commented on src/irr/asset/IAssetManager.cpp:
```diff
@@ -90,7 +90,7 @@ std::function<void(SAssetBundle&)> irr::asset::makeAssetGreetFunc(const IAssetMa
 		//assets being in the cache must be immutable
         for (auto ass : rng)
-			_mgr->setAssetMutable(ass.get(), false);
+			_mgr->setAssetMutability(ass.get(), IAsset::EM_IMMUTABLE);
```
why are you changing our default caching to "always immutable" ?

devshgraphicsprogramming replied:
default behaviour should be no change (assets go in as MUTABLE)

Crisspl committed:
Immutable assets: fixed convertToDummy(), dummy restoration, changes in default mutability levels

Builtins are EM_IMMUTABLE, not-builtins which are kept in cache are EM_CPU_PERSISTENT, everything else is EM_MUTABLE
TODO: assets being kept in cache (beside builtins) should be EM_MUTABLE, only assets explicitely cached by user should be EM_CPU_PERSISTENT

Crisspl committed:
Resolved TODO from prev commit

devshgraphicsprogramming commented on include/irr/asset/IAsset.h:
```diff
@@ -173,6 +173,17 @@ class IAsset : virtual public core::IReferenceCounted
+		{
+			if (!this->isADummyObjectForCache())
+				return false;
+			assert(!_other->isADummyObjectForCache());
```
would be useful if we called it `swapContents` and had the two assets swap stuff around

devshgraphicsprogramming commented on include/irr/asset/IAssetManager.h:
```diff
@@ -422,8 +422,7 @@ class IAssetManager : public core::IReferenceCounted, public core::QuitSignallin
         {
-            const uint32_t ix = IAsset::typeFlagToIndex(_asset.getAssetType());
-            return m_assetCache[ix]->insert(_asset.getCacheKey(), _asset);
+            return insertAssetIntoCache(_asset, IAsset::EM_CPU_PERSISTENT);
```
this is now mutable by default, yes?

Crisspl replied:
public `IAssetManager::insertIntoCache` sets asset's mutability as cpu persistent (so that everything that is added to cache explicitely by user is cpu persistent), but `IAssetLoaderOverride` sets mutability as mutable (everything loaded by `getAsset` and subassets that are added to cache during the loading process are mutable)
All the builtins are added to cache as immutable

devshgraphicsprogramming replied:
ok I guess it kind-of makes sense.

devshgraphicsprogramming replied:
not entirely convinced about the asset manager default behaviour.

devshgraphicsprogramming replied:
hmm ok can keep as is, just document want Criss wrote on the wiki @AnastaZIuk

devshgraphicsprogramming replied:
but I'd instead add a `mutability` argument to the function with ` IAsset::EM_CPU_PERSISTENT` as the default parameter so user doesn't have to do the long form of first inserting an asset and then changing its mutability to what he actually wants

devshgraphicsprogramming commented on include/irr/asset/ICPUPipelineCache.h:
```diff
@@ -110,10 +110,21 @@ class ICPUPipelineCache final : public IAsset
+			return false;
+
+		auto* other = static_cast<ICPUPipelineCache*>(_other);
+		std::swap(m_cache, other->m_cache);
```
can we even restore this?

Crisspl replied:
hm.. good point. This asset is kinda special because the idea was GPU object->asset->potentially save on disk

devshgraphicsprogramming replied:
we'd have to check a few things before restoring, then I guess?

devshgraphicsprogramming commented on include/irr/asset/ICPUBuffer.h:
```diff
@@ -60,6 +60,18 @@ class ICPUBuffer : public asset::IBuffer, public asset::IAsset
+
         virtual void convertToDummyObject(uint32_t referenceLevelsBelowToConvert=0u) override
         {
             if (isDummyObjectForCacheAliasing)
```
maybe fold this check into `canBeConvertedToDummy`

devshgraphicsprogramming replied:
btw, `convertToDummyObject_common` will set `isDummyObjectForCacheAliasing = true;`

devshgraphicsprogramming commented on include/irr/asset/ICPUPipelineCache.h:
```diff
@@ -110,10 +110,21 @@ class ICPUPipelineCache final : public IAsset
 	void convertToDummyObject(uint32_t referenceLevelsBelowToConvert = 0u) override
 	{
-		if (canBeConvertedToDummy())
+		if (!isADummyObjectForCache() && canBeConvertedToDummy())
```
as I wrote, you could fold `usADummyObjectForCache` into `canBeConvertedToDummy`

Crisspl committed:
Immutable assets: new stuff talked over on dc

Crisspl committed:
A few small fixes regarding immutable assets

devshgraphicsprogramming commented on include/irr/asset/IAsset.h:
```diff
@@ -196,7 +196,7 @@ class IAsset : virtual public core::IReferenceCounted
 		inline E_MUTABILITY getMutability() const { return m_mutability; }
 		inline bool isMutable() const { return getMutability() == EM_MUTABLE; }
-		inline bool canBeConvertedToDummy() const { return (getMutability()&EM_CPU_PERSISTENT) != EM_CPU_PERSISTENT; }
+		inline bool canBeConvertedToDummy() const { return !isADummyObjectForCache() && (getMutability()&EM_CPU_PERSISTENT) != EM_CPU_PERSISTENT; }
```
maybe `getMutability()<EM_CPU_PERSISTENT`

devshgraphicsprogramming replied:
instead of bitfield

devshgraphicsprogramming replied:
TODO

devshgraphicsprogramming commented on include/irr/asset/IAsset.h:
```diff
@@ -173,14 +173,24 @@ class IAsset : virtual public core::IReferenceCounted
 
-		//! restore() is not recursive!
-		virtual bool restore(IAsset* _other)
+		bool restoreFromDummy(IAsset* _other, uint32_t _levelsBelow)
```
needs default argument for `_levelsBelow` should `~0u`

devshgraphicsprogramming commented on include/irr/asset/IAsset.h:
```diff
@@ -173,14 +173,24 @@ class IAsset : virtual public core::IReferenceCounted
+			if (getMutability() != EM_MUTABLE)
+				return false;
+			if (!isADummyObjectForCache())
+				return false;
```
this should not be here, just because asset is not mutable or is not a dummy object (cannot be restored, doesn't matter what `_other` is) does not meen the dependencies couldn't be restored.

Same if `_other` is a dummy object, the dependants of `this` which are mutable and dummy could be restored from dependants of `_other` which are mutable and non-dummy.

**However if `_other` does not "match"/"is not compatible" with `this` (i.e. different layout contents) then obviously there's no point restoring the dependants.** 

P.S. when restoring you should check that `_other` is mutable.

devshgraphicsprogramming commented on include/irr/asset/IAsset.h:
```diff
@@ -173,14 +173,24 @@ class IAsset : virtual public core::IReferenceCounted
+			if (!canBeRestoredFrom(_other))
+				return false;
+
+			restoreFromDummy_impl(_other, _levelsBelow);
```
you should probably split this into `restoreFromDummy_this` and `restoreFromDummy_recurse`, to solve comment
https://github.com/buildaworldnet/IrrlichtBAW/pull/549/commits/750f5f9dfa92c3d1e05482fca38a5723768a4682#r507655849

you can only restore `this` from `_other` if:
1. `_other` matches `this`
2. `this` is dummy
3. `this` is mutable
4. `_other` is not a dummy
5. `_other` is mutable

You can only recurse if **`_other` "matches" `_this`, no other conditions necessary**

devshgraphicsprogramming commented on include/irr/asset/ICPUBuffer.h:
```diff
@@ -112,6 +112,13 @@ class ICPUBuffer : public asset::IBuffer, public asset::IAsset
+        {
+            auto* other = static_cast<ICPUBuffer*>(_other);
+
+            std::swap(data, other->data);
```
you're forgetting the size.

devshgraphicsprogramming replied:
ah ok, size is supposed to be identical

devshgraphicsprogramming commented on include/irr/asset/ICPUComputePipeline.h:
```diff
@@ -32,9 +32,9 @@ class ICPUComputePipeline : public IComputePipeline<ICPUSpecializedShader, ICPUP
-        if (!IAsset::restore(_other))
+        if (!IAsset::canBeRestoredFrom(_other))
             return false;
 
```
the "does `this` match `_other`"restoration check needs to be fully recursive, so need to check whether layouts match (etc.) before trying to restore top level objects

devshgraphicsprogramming replied:
kinda paired up with https://github.com/buildaworldnet/IrrlichtBAW/pull/549#discussion_r507659230

devshgraphicsprogramming commented on include/irr/asset/ICPUDescriptorSet.h:
```diff
@@ -193,6 +193,41 @@ class ICPUDescriptorSet final : public IDescriptorSet<ICPUDescriptorSetLayout>,
+							it->image.sampler->restoreFromDummy(other->m_descriptors->begin()[i].image.sampler.get(), _levelsBelow);
+						break;
+					case IDescriptor::EC_BUFFER_VIEW:
+						static_cast<asset::ICPUBufferView*>(descriptor)->restoreFromDummy(static_cast<ICPUBufferView*>(d_other), _levelsBelow);
```
if we'll already check if we can restore entire dependency tree, then this should call `restoreFromDummy_impl` directly instead of `restoreFromDummy` for dependencies

Tied up with the same thread https://github.com/buildaworldnet/IrrlichtBAW/pull/549#discussion_r507664092

devshgraphicsprogramming commented on include/irr/asset/ICPUDescriptorSetLayout.h:
```diff
@@ -66,21 +66,29 @@ class ICPUDescriptorSetLayout : public IDescriptorSetLayout<ICPUSampler>, public
+
+        //TODO hmm should samplers be swapped (and dropped in convertToDummy()) or recursively restored??
+        std::swap(m_bindings, other->m_bindings);
+        std::swap(m_samplers, other->m_samplers);
```
pointer values of `m_samplers` should not be swapped, only contents (because it would mess up CPU 2 GPU caches)

so recursively restore.

devshgraphicsprogramming commented on include/irr/asset/ICPUImage.h:
```diff
@@ -206,6 +206,14 @@ class ICPUImage final : public IImage, public IAsset
+		{
+			auto* other = static_cast<ICPUImage*>(_other);
+
+			std::swap(regions, other->regions);
```
there's also a buffer to consider, not only regions

devshgraphicsprogramming commented on include/irr/asset/ICPUShader.h:
```diff
@@ -55,9 +55,13 @@ class ICPUShader : public IAsset, public IShader
+				return false;
+
+			auto* other = static_cast<const ICPUShader*>(_other);
+			if (m_containsGLSL != other->m_containsGLSL)
```
need to think about whether to require this.

devshgraphicsprogramming commented on include/irr/asset/ICPUSpecializedShader.h:
```diff
@@ -65,17 +65,18 @@ class ICPUSpecializedShader : public IAsset, public ISpecializedShader
+			auto* other = static_cast<const ICPUSpecializedShader*>(_other);
+			if (m_specInfo.entryPoint != other->m_specInfo.entryPoint)
+				return false;
+			if (m_specInfo.m_filePathHint != other->m_specInfo.m_filePathHint)
```
should we require this?

devshgraphicsprogramming replied:
we probably shouldn't require this.

devshgraphicsprogramming commented on include/irr/asset/IAssetManager.h:
```diff
@@ -666,7 +666,7 @@ class IAssetManager : public core::IReferenceCounted, public core::QuitSignallin
         }
 
-        void restoreDummyAsset(SAssetBundle& _bundle)
+        void restoreDummyAsset(SAssetBundle& _bundle, uint32_t _levelsBelow = 0u)
```
this function is dangerous, it takes no limits/constraints on asset type to be restored (you will usually know the type of asset you're after).

The amount of duplication should be not top level, but top level + `_levelsBelow`

Also handling of mixed dummy and non dummy assets in a bundle is "meh".

Also no check that length of new bundle == old bundle

Crisspl replied:
> this function is dangerous, it takes no limits/constraints on asset type to be restored

asset bundle can hold only one type of assets if thats what you mean

devshgraphicsprogramming replied:
no that's not what I meant, I mean that you pass a bundle of some type, then use the bundle's cache key value to try and get a fresh load (badly because only top level will be duplicated and for restore you need the full DAG).

Then you dont limit the asset types you will retrieve/load from `getAssetInHierarchy` so the bundle you might get back could be a different length and even type.

Furthermore you `assert` and not fail gracefully if you dont get same size bundle.

Then you restore without extra checks.

Its basically a surefire way to program termination either via `assert` or trying to restore assets from different types.

Basically the usefulness of this method approaches 0

Crisspl committed:
tmp

Crisspl committed:
Immutable assets: reworked (talked through over dc)

devshgraphicsprogramming commented on include/irr/asset/IAsset.h:
```diff
@@ -175,17 +175,23 @@ class IAsset : virtual public core::IReferenceCounted
+				return false;
 			if (!isADummyObjectForCache())
 				return false;
 			if (_other->isADummyObjectForCache())
```
needs a call to `canBeRestoredFrom_recurseDAG` at the end

devshgraphicsprogramming replied:
but the `restoreFromDummy` should not call `canBeRestoredFrom` but `canBeRestoredFrom_this_node` or something like that

Crisspl replied:
> needs a call to canBeRestoredFrom_recurseDAG at the end

what, no, why

Crisspl replied:
what i've noted from our conversation:
* `restoreFromDummy` first calls `canBeRestoredFrom_recurseDAG` and early exits if false, otherwise calls `restoreFromDummy_impl`
* `restoreFromDummy_impl` calls `canBeRestoredFrom` and restores if true, but regardless of result calls `restoreFromDummy` on dependencies

devshgraphicsprogramming replied:
What you're missing that is the public `canBeRestoredFrom` checks for whether mutability and dummy state of self and this is ok (what I suggest calling `canBeRestoredFrom_this_node`, which would be equivalent to your current 4 if-statements) AND THEN the "matching" of the DAG

because of this, the `restoreFromDummy_impl` should not call this function to prevent the resotration of current asset in the DAG, but a protected `canBeRestoredFrom_this_node`.

`restoreFromDummy` shouldn't call the public function either but only `canBeRestoredFrom_recurseDAG` to early out.

Crisspl replied:
soo
> needs a call to canBeRestoredFrom_recurseDAG at the end

is not the case
Your description above perfectly matches what actually is now

devshgraphicsprogramming replied:
no its not, there's one thing missing

The fucntion here, `canBeRestoredFrom`, should do the 4 if statements plus a call to `canBeRestoredFrom_recurseDAG`

devshgraphicsprogramming replied:
ugh I said it wrong again.

devshgraphicsprogramming commented on include/irr/asset/IAsset.h:
```diff
@@ -194,6 +200,8 @@ class IAsset : virtual public core::IReferenceCounted
 			return true;
 		}
 
+		virtual bool canBeRestoredFrom_recurseDAG(const IAsset* _other) const = 0;
```
make it protected

Crisspl replied:
cant because an asset has to be able to call it on its dependencies

devshgraphicsprogramming replied:
but asset and its dependencies both inherit from `IAsset` ?

Crisspl replied:
doesnt matter

devshgraphicsprogramming replied:
ok there's a simple workaround for that, move `canBeRestoredFrom_recurseDAG` to protected and also provide this as a workaround:
```
inline bool canBeRestoredFrom_recurseDAG_call(const IAsset* _other) const
{
    return canBeRestoredFrom_recurseDAG(_other);
}
```
then because `canBeRestoredFrom_recurseDAG_call` is not virtual derived classes should be able to call it, correcT?

Crisspl replied:
> then because canBeRestoredFrom_recurseDAG_call is not virtual derived classes should be able to call it, correcT?

no, why?
hm, i think workaround would be put this into IAsset
```
inline static bool IAsset::canBeRestoredFrom_recurseDAG_call(const IAsset* _this_child, const IAsset* _other_child)
{
    return _this_child->canBeRestoredFrom_recurseDAG(_other_child);
}
```
and then call in derived classes, i.e. in case of DS:
`canBeRestoredFrom_recurseDAG_call(getLayout(), other->getLayout())`

same for `restoreFromDummy_impl`

Crisspl replied:
im doing this

devshgraphicsprogramming commented on include/irr/asset/ICPUDescriptorSetLayout.h:
```diff
@@ -85,9 +82,14 @@ class ICPUDescriptorSetLayout : public IDescriptorSetLayout<ICPUSampler>, public
+        {
+            std::swap(m_bindings, other->m_bindings);
+            std::swap(m_samplers, other->m_samplers);
+        }
```
you should recursively restore, dont swap what are essentially pointer lists!

devshgraphicsprogramming commented on include/irr/asset/ICPUDescriptorSetLayout.h:
```diff
@@ -66,11 +66,8 @@ class ICPUDescriptorSetLayout : public IDescriptorSetLayout<ICPUSampler>, public
         {
-            if (!IAsset::canBeRestoredFrom(_other))
-                return false;
-
```
see https://github.com/buildaworldnet/IrrlichtBAW/pull/549/files/e94f07ab352808077b42c37c490b5c93986df2e5..2ddd5e994ba249c82b43de1b43625f1ff57465a3#r508847179

devshgraphicsprogramming commented on include/irr/asset/ICPURenderpassIndependentPipeline.h:
```diff
@@ -55,6 +52,16 @@ class ICPURenderpassIndependentPipeline : public IRenderpassIndependentPipeline<
 
+			for (uint32_t i = 0u; i < SHADER_STAGE_COUNT; ++i)
+			{
+				if ((!m_shaders[i] || !other->m_shaders[i]) && m_shaders[i] != other->m_shaders[i])
```
I dont get this if statement

devshgraphicsprogramming commented on include/irr/asset/ICPUPipelineLayout.h:
```diff
@@ -58,11 +58,25 @@ class ICPUPipelineLayout : public IAsset, public IPipelineLayout<ICPUDescriptorS
-                return false;
+            auto* other = static_cast<const ICPUPipelineLayout*>(_other);
 
+            if ((!m_pushConstantRanges || !other->m_pushConstantRanges) && m_pushConstantRanges != other->m_pushConstantRanges)
```
I dont get this if statement

devshgraphicsprogramming commented on include/irr/asset/ICPUPipelineLayout.h:
```diff
@@ -58,11 +58,25 @@ class ICPUPipelineLayout : public IAsset, public IPipelineLayout<ICPUDescriptorS
+            if ((!m_pushConstantRanges || !other->m_pushConstantRanges) && m_pushConstantRanges != other->m_pushConstantRanges)
+                return false;
+            if (m_pushConstantRanges->size() != other->m_pushConstantRanges->size())
+                return false;
```
this is bad, its okay to have no push constant ranges, this will fail if either is 0

devshgraphicsprogramming commented on include/irr/asset/ICPUPipelineLayout.h:
```diff
@@ -58,11 +58,25 @@ class ICPUPipelineLayout : public IAsset, public IPipelineLayout<ICPUDescriptorS
+
+            for (uint32_t i = 0u; i < DESCRIPTOR_SET_COUNT; ++i)
+            {
+                if ((!m_descSetLayouts[i] || !other->m_descSetLayouts[i]) && m_descSetLayouts[i] != other->m_descSetLayouts[i])
```
I dont get this if statement

devshgraphicsprogramming replied:
missed this one

devshgraphicsprogramming commented on include/irr/asset/ICPUMeshBuffer.h:
```diff
@@ -149,6 +146,31 @@ class ICPUMeshBuffer : public IMeshBuffer<ICPUBuffer, ICPUDescriptorSet, ICPURen
             return false;
+        if (m_indexBufferBinding.offset != other->m_indexBufferBinding.offset)
+            return false;
+        if ((!m_indexBufferBinding.buffer || !other->m_indexBufferBinding.buffer) && m_indexBufferBinding.buffer != other->m_indexBufferBinding.buffer)
```
I dont get this if statement

devshgraphicsprogramming commented on include/irr/asset/ICPUMeshBuffer.h:
```diff
@@ -149,6 +146,31 @@ class ICPUMeshBuffer : public IMeshBuffer<ICPUBuffer, ICPUDescriptorSet, ICPURen
+            return false;
+        if ((!m_indexBufferBinding.buffer || !other->m_indexBufferBinding.buffer) && m_indexBufferBinding.buffer != other->m_indexBufferBinding.buffer)
+            return false;
+        if (!m_indexBufferBinding.buffer->canBeRestoredFrom_recurseDAG(other->m_indexBufferBinding.buffer.get()))
```
will fail with null index buffers

devshgraphicsprogramming commented on include/irr/asset/ICPUMeshBuffer.h:
```diff
@@ -149,6 +146,31 @@ class ICPUMeshBuffer : public IMeshBuffer<ICPUBuffer, ICPUDescriptorSet, ICPURen
+                return false;
+            if ((!m_vertexBufferBindings[i].buffer || !other->m_vertexBufferBindings[i].buffer) && m_vertexBufferBindings[i].buffer != other->m_vertexBufferBindings[i].buffer)
+                return false;
+            if (!m_vertexBufferBindings[i].buffer->canBeRestoredFrom_recurseDAG(other->m_vertexBufferBindings[i].buffer.get()))
```
same bug as all your optional pointer comparisons

devshgraphicsprogramming commented on include/irr/asset/ICPUMeshBuffer.h:
```diff
@@ -149,6 +146,31 @@ class ICPUMeshBuffer : public IMeshBuffer<ICPUBuffer, ICPUDescriptorSet, ICPURen
+
+        if ((!m_descriptorSet || !other->m_descriptorSet) && m_descriptorSet != other->m_descriptorSet)
+            return false;
+        if (!m_descriptorSet->canBeRestoredFrom_recurseDAG(other->m_descriptorSet.get()))
```
pointer mess again

devshgraphicsprogramming commented on include/irr/asset/ICPUMeshBuffer.h:
```diff
@@ -149,6 +146,31 @@ class ICPUMeshBuffer : public IMeshBuffer<ICPUBuffer, ICPUDescriptorSet, ICPURen
+
+        if ((!m_pipeline || !other->m_pipeline) && m_pipeline != other->m_pipeline)
+            return false;
+        if (!m_pipeline->canBeRestoredFrom_recurseDAG(other->m_pipeline.get()))
```
pipeline is not optional IIRC

Crisspl replied:
it is, it's not used at all right now in a way other than returning pipeline from mesh loader

devshgraphicsprogramming replied:
you cant draw a meshbuffer without a pipeline, you dont even know what primitive type it is

Crisspl replied:
i mean `video` namespace ignores it

devshgraphicsprogramming replied:
!?

AFAIK its an error to have a ICPUMeshBuffer with a nullptr ICPUPipeline

Crisspl replied:
never heard of this really

Crisspl committed:
Fixed if statements in canBeRestoredFrom_recurseDAG()

Crisspl committed:
Immutable assets: cpu2gpu with const ptrs + some fixes

Crisspl committed:
Resolved comments from imm assets PR

Crisspl committed:
Made new type traits for pointers preserve constness of each ptr level (for cpu2gpu)

devshgraphicsprogramming commented on include/IDriver.h:
```diff
@@ -376,7 +376,7 @@ class IDriver : public virtual core::IReferenceCounted, public IVideoCapabilityR
         // With a custom converter, you can override it to for example; pack all buffers into one, pack all images into one atlas, etc.
 
-        template<typename AssetType>
+        /*template<typename AssetType>
```
what happened here?

devshgraphicsprogramming commented on include/irr/asset/CCPUMesh.h:
```diff
@@ -55,15 +55,23 @@ class CCPUMesh final : public ICPUMesh
+		virtual ICPUMeshBuffer* getMeshBuffer(uint32_t nr) override
+		{
+			if (isImmutable_debug())
+				return nullptr;
```
just assert ;)

devshgraphicsprogramming commented on include/irr/asset/IAsset.h:
```diff
@@ -206,8 +206,8 @@ class IAsset : virtual public core::IReferenceCounted
 		inline bool canBeConvertedToDummy() const { return !isADummyObjectForCache() && (getMutability()&EM_CPU_PERSISTENT) != EM_CPU_PERSISTENT; }
 
-    protected:
 		virtual void restoreFromDummy_impl(IAsset* _other, uint32_t _levelsBelow) = 0;
```
why public?

devshgraphicsprogramming commented on include/irr/asset/ICPUBuffer.h:
```diff
@@ -99,22 +99,21 @@ class ICPUBuffer : public asset::IBuffer, public asset::IAsset
-        uint64_t size;
-        void* data;
-
         void restoreFromDummy_impl(IAsset* _other, uint32_t _levelsBelow) override
```
why public?

devshgraphicsprogramming commented on include/irr/asset/ICPUComputePipeline.h:
```diff
@@ -58,32 +58,32 @@ class ICPUComputePipeline : public IComputePipeline<ICPUSpecializedShader, ICPUP
         m_shader = core::smart_refctd_ptr<ICPUSpecializedShader>(_cs); 
     }
 
-private:
```
why

devshgraphicsprogramming commented on include/irr/asset/ICPUPipelineCache.h:
```diff
@@ -133,7 +133,6 @@ class ICPUPipelineCache final : public IAsset
 		return core::make_smart_refctd_ptr<ICPUPipelineCache>(std::move(cache_cp));
 	}
 
-private:
```
why

devshgraphicsprogramming commented on source/Irrlicht/IDriver.cpp:
```diff
@@ -57,7 +57,7 @@ struct AssetBundleIterator
 
 //! Maybe we can reduce code duplication here some day
-template<typename AssetType>
+/*template<typename AssetType>
```
I need an explainer about this

Crisspl replied:
ah i forgot to re-enable this

devshgraphicsprogramming commented on src/irr/asset/CPLYMeshWriter.cpp:
```diff
@@ -169,8 +169,10 @@ bool CPLYMeshWriter::writeAsset(io::IWriteFile* _file, const SAssetWriteParams&
-    else
-        indices = meshBuffer->getIndices();
+    else {
+        indices = _IRR_ALIGNED_MALLOC(idxCnt * (idxtype == asset::EIT_16BIT ? 2u : 4u), _IRR_SIMD_ALIGNMENT);
+        memcpy(indices, meshBuffer->getIndices(), idxCnt * (idxtype == asset::EIT_16BIT ? 2u : 4u));
```
why?

Crisspl committed:
Re-enabled AssetBundleIterator stuff in IDriver

Crisspl committed:
Immutable assets: workarounds for hermetization

devshgraphicsprogramming commented on include/irr/asset/IAsset.h:
```diff
@@ -167,25 +173,81 @@ class IAsset : virtual public core::IReferenceCounted
+			_this_child->restoreFromDummy_impl(_other_child, _levelsBelow);
+		}
+
+		virtual bool canBeRestoredFrom_recurseDAG(const IAsset* _other) const = 0;
```
should take a `uint32_t _levelsBelow` parameter

devshgraphicsprogramming replied:
oh, we already discussed it

devshgraphicsprogramming replied:
just remind me of the corner case that would break a `_levelsBelow` parameter version?

Crisspl replied:
anything, you want to restore pipeline to the level of pipeline layout so you check only those levels but it might turn out that pipelines are incompatible on some lower level (ds layouts for example)

devshgraphicsprogramming replied:
ok, thx

devshgraphicsprogramming commented on include/irr/asset/ICPUImageView.h:
```diff
@@ -57,9 +55,44 @@ class ICPUImageView final : public IImageView<ICPUImage>, public IAsset
+			auto* other = static_cast<const ICPUImageView*>(_other);
+			const auto& rhs = other->params;
+
+			if (params.flags != rhs.flags)
+				return false;
+			if (params.format != rhs.format)
+				return false;
+			if (params.components != rhs.components)
+				return false;
+			if (params.viewType != rhs.viewType)
+				return false;
+			if (memcmp(&params.subresourceRange, &rhs.subresourceRange, sizeof(params.subresourceRange)))
+				return false;
```
ugh this is so coupled.

Crisspl committed:
Finished immutable assets PR

Crisspl committed:
Merge branch 'property_pool' into 3rdparty-upgrade
